File: app/src/main/java/ly/pp/justpiano3/adapter/LocalSongsAdapter.java
package ly.pp.justpiano3.adapter;

import android.content.Context;
import android.database.Cursor;
import android.text.method.ScrollingMovementMethod;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CursorAdapter;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;
import ly.pp.justpiano3.R;
import ly.pp.justpiano3.activity.MelodySelect;
import ly.pp.justpiano3.listener.AddSongsFavoClick;
import ly.pp.justpiano3.listener.LocalSongsStartPlayClick;
import ly.pp.justpiano3.listener.PlaySongsClick;
import ly.pp.justpiano3.view.ScrollText;

public final class LocalSongsAdapter extends CursorAdapter {
    public final MelodySelect melodyselect;

    public LocalSongsAdapter(MelodySelect melodySelect, Context context, Cursor cursor) {
        super(context, cursor, true);
        melodyselect = melodySelect;
    }

    public static void m3966a(LocalSongsAdapter LocalSongsAdapter, int i) {
        switch (i) {
            case 0:
                LocalSongsAdapter.melodyselect.mo2784a(LocalSongsAdapter.melodyselect.sqlitedatabase.query("jp_data", null, "isfavo = 1 AND " + LocalSongsAdapter.melodyselect.f4238O + " AND ishot=0 AND " + LocalSongsAdapter.melodyselect.f4238O, null, null, null, LocalSongsAdapter.melodyselect.sortStr));
                return;
            case 1:
            case 4:
            case 2:
            case 6:
            case 5:
            case 3:
                LocalSongsAdapter.melodyselect.mo2784a(LocalSongsAdapter.melodyselect.sqlitedatabase.query("jp_data", null, "ishot = 0 AND " + LocalSongsAdapter.melodyselect.f4238O + " AND item=?", new String[]{LocalSongsAdapter.melodyselect.f4263s}, null, null, LocalSongsAdapter.melodyselect.sortStr));
                return;
            case 7:
                LocalSongsAdapter.melodyselect.mo2784a(LocalSongsAdapter.melodyselect.sqlitedatabase.query("jp_data", null, "ishot = 1 AND " + LocalSongsAdapter.melodyselect.f4238O + " AND isfavo=0", new String[]{LocalSongsAdapter.melodyselect.f4263s}, null, null, LocalSongsAdapter.melodyselect.sortStr));
                return;
            default:
        }
    }

    @Override
    public void bindView(View view, Context context, Cursor cursor) {
        view.setTag(cursor.getPosition());
        ImageView imageView = view.findViewById(R.id.play_image);
        ImageView imageView3 = view.findViewById(R.id.favor);
        view.findViewById(R.id.play).setOnClickListener(new LocalSongsStartPlayClick(this, cursor));
        imageView.setOnClickListener(new PlaySongsClick(this, cursor, view, context));
        imageView3.setOnClickListener(new AddSongsFavoClick(this, cursor, context));
        ScrollText scrollText = view.findViewById(R.id.s_n);
        scrollText.setText(cursor.getString(1));
        scrollText.setMovementMethod(ScrollingMovementMethod.getInstance());
        scrollText.setHorizontallyScrolling(true);
        TextView textView = view.findViewById(R.id.highscore);
        scrollText.setOnClickListener(new LocalSongsStartPlayClick(this, cursor));
        int i = cursor.getInt(cursor.getColumnIndexOrThrow("score"));
        int i2 = cursor.getInt(cursor.getColumnIndexOrThrow("isnew"));
        textView.setText("最高分: " + (i <= 0 ? "0" : String.valueOf(i)));
        RatingBar ratingBar = view.findViewById(R.id.nandu);
        textView = view.findViewById(R.id.sound_time);
        int l = cursor.getInt(14);
        String str1 = l / 60 >= 10 ? "" + l / 60 : "0" + l / 60;
        String str2 = l % 60 >= 10 ? "" + l % 60 : "0" + l % 60;
        textView.setText(str1 + ":" + str2);
        textView = view.findViewById(R.id.nandu_1);
        float f = cursor.getFloat(11);
        if ((int) f == 10) {
            textView.setText(" 难度: 右手 10 ");
        } else {
            textView.setText(" 难度: 右手 " + f);
        }
        ratingBar.setNumStars(5);
        ratingBar.setClickable(false);
        ratingBar.setRating(f / 2);
        RatingBar ratingBar2 = view.findViewById(R.id.leftnandu);
        textView = view.findViewById(R.id.nandu_2);
        float j = cursor.getFloat(13);
        if ((int) j == 10) {
            textView.setText(" 左手 10");
        } else {
            textView.setText(" 左手 " + j);
        }
        ratingBar2.setNumStars(5);
        ratingBar2.setClickable(false);
        ratingBar2.setRating(j / 2);
        melodyselect.f4228E = view.findViewById(R.id.is_new);
        if (i2 == 1) {
            melodyselect.f4228E.setImageResource(R.drawable.s_new);
        } else {
            melodyselect.f4228E.setImageResource(R.drawable.null_pic);
        }
    }

    @Override
    public View newView(Context context, Cursor cursor, ViewGroup viewGroup) {
        View inflate = melodyselect.layoutInflater2.inflate(R.layout.c_view, null);
        inflate.setBackgroundResource(R.drawable.selector_list_c);
        return inflate;
    }
}

File: app/src/main/java/ly/pp/justpiano3/adapter/LocalSongsAdapter.kt
package ly.pp.justpiano3.adapter

import android.content.Intent
import android.text.method.ScrollingMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.TextView
import android.widget.Toast
import androidx.paging.PagedList
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.RecyclerView
import ly.pp.justpiano3.JPApplication
import ly.pp.justpiano3.R
import ly.pp.justpiano3.activity.local.MelodySelect
import ly.pp.justpiano3.activity.local.WaterfallActivity
import ly.pp.justpiano3.constant.Consts
import ly.pp.justpiano3.database.entity.Song
import ly.pp.justpiano3.listener.LocalSongsStartPlayClick
import ly.pp.justpiano3.thread.SongPlay
import ly.pp.justpiano3.view.ScrollText
import java.io.File

class LocalSongsAdapter(
    private val melodySelect: MelodySelect,
    private val songsListView: RecyclerView
) :
    PagedListAdapter<Song, LocalSongsAdapter.SongViewHolder>(Consts.SONG_DIFF_UTIL) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SongViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val itemView = inflater.inflate(R.layout.lo_songview, parent, false)
        itemView.setBackgroundResource(R.drawable.selector_list_c)
        return SongViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: SongViewHolder, position: Int) {
        // 绑定数据到ViewHolder
        getItem(position)?.let { holder.bindData(position, it) }
    }

    override fun onCurrentListChanged(
        previousList: PagedList<Song>?,
        currentList: PagedList<Song>?
    ) {
        super.onCurrentListChanged(previousList, currentList)
        // 当数据列表发生更改时，滚动到第一个位置
        songsListView.scrollToPosition(0)
    }

    fun getSongByPosition(position: Int): Song? {
        return getItem(position)
    }

    inner class SongViewHolder(songView: View) : RecyclerView.ViewHolder(songView) {
        private val listenImageView: ImageView
        private val waterFallImageView: ImageView
        private val favorImageView: ImageView
        private val isNewImageView: ImageView
        private val playImageView: ImageView
        private val highScoreTextView: TextView
        private val soundTimeTextView: TextView
        private val rightHandDegreeTextView: TextView
        private val leftHandDegreeTextView: TextView
        private val rightHandDegreeRatingBar: RatingBar
        private val leftHandDegreeRatingBar: RatingBar
        private val songNameScrollText: ScrollText

        fun bindData(position: Int, song: Song) {
            playImageView.setOnClickListener(LocalSongsStartPlayClick(melodySelect, song))
            listenImageView.setOnClickListener {
                if (song.filePath == melodySelect.songsPath && SongPlay.isPlaying()) {
                    melodySelect.songsPath = ""
                    SongPlay.stopPlay()
                    return@setOnClickListener
                }
                melodySelect.songsPath = song.filePath
                melodySelect.songsPositionInListView = position
                SongPlay.startPlay(melodySelect, song.filePath, 0)
                Toast.makeText(melodySelect, "开始播放:《" + song.name + "》", Toast.LENGTH_SHORT)
                    .show()
            }
            waterFallImageView.setOnClickListener {
                SongPlay.stopPlay()
                val intent = Intent()
                intent.putExtra("songPath", song.filePath)
                intent.setClass(melodySelect, WaterfallActivity::class.java)
                melodySelect.startActivity(intent)
            }
            if (song.category == Consts.items[Consts.items.size - 1]) {
                favorImageView.setImageResource(R.drawable.dele)
                favorImageView.setOnClickListener {
                    val songDao = JPApplication.getSongDatabase().songDao()
                    songDao.deleteSongs(listOf(song))
                    File(melodySelect.filesDir.absolutePath + "/ImportSongs/" + song.name + ".pm").delete()
                    Toast.makeText(melodySelect, song.name + ":已删除", Toast.LENGTH_SHORT).show()
                }
            } else {
                favorImageView.setImageResource(if (song.isFavorite == 0) R.drawable.favor_1 else R.drawable.favor)
                favorImageView.setOnClickListener {
                    val songDao = JPApplication.getSongDatabase().songDao()
                    melodySelect.pagedListLiveData.removeObservers(melodySelect)
                    songDao.updateFavoriteSong(song.filePath, if (song.isFavorite == 0) 1 else 0)
                    val songsDataSource =
                        songDao.getLocalSongsWithDataSource(0, melodySelect.orderPosition)
                    melodySelect.pagedListLiveData =
                        songDao.getPageListByDatasourceFactory(songsDataSource)
                    melodySelect.pagedListLiveData.observe(melodySelect) { pagedList: PagedList<Song>? ->
                        this@LocalSongsAdapter.submitList(pagedList)
                    }
                    Toast.makeText(
                        melodySelect,
                        song.name + if (song.isFavorite == 0) ":已加入收藏夹" else ":已移出收藏夹",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
            songNameScrollText.text = song.name
            songNameScrollText.movementMethod = ScrollingMovementMethod.getInstance()
            songNameScrollText.setHorizontallyScrolling(true)
            songNameScrollText.setOnClickListener(LocalSongsStartPlayClick(melodySelect, song))
            val rightHandScore = song.rightHandHighScore
            val leftHandScore = song.leftHandHighScore
            highScoreTextView.text =
                "右手最高:" + (if (rightHandScore <= 0) "0" else rightHandScore.toString()) +
                        " 左手最高: " + if (leftHandScore <= 0) "0" else leftHandScore.toString()
            val length = song.length
            val str1 = if (length / 60 >= 10) "" + length / 60 else "0" + length / 60
            val str2 = if (length % 60 >= 10) "" + length % 60 else "0" + length % 60
            soundTimeTextView.text = "$str1:$str2"
            val rightHandDegree = song.rightHandDegree
            rightHandDegreeTextView.text =
                if (rightHandDegree.toInt() == 10) " 难度: 右手 10 " else " 难度: 右手 $rightHandDegree"
            rightHandDegreeRatingBar.numStars = 5
            rightHandDegreeRatingBar.isClickable = false
            rightHandDegreeRatingBar.rating = rightHandDegree / 2
            val leftHandDegree = song.leftHandDegree
            leftHandDegreeTextView.text =
                if (leftHandDegree.toInt() == 10) " 左手 10" else " 左手 $leftHandDegree"
            leftHandDegreeRatingBar.numStars = 5
            leftHandDegreeRatingBar.isClickable = false
            leftHandDegreeRatingBar.rating = leftHandDegree / 2
            isNewImageView.setImageResource(if (song.isNew == 1) R.drawable.s_new else R.drawable.null_pic)
        }

        init {
            listenImageView = songView.findViewById(R.id.play_image)
            waterFallImageView = songView.findViewById(R.id.play_waterfall)
            favorImageView = songView.findViewById(R.id.favor)
            isNewImageView = songView.findViewById(R.id.is_new)
            playImageView = songView.findViewById(R.id.play)
            highScoreTextView = songView.findViewById(R.id.highscore)
            soundTimeTextView = songView.findViewById(R.id.sound_time)
            rightHandDegreeTextView = songView.findViewById(R.id.nandu_1)
            leftHandDegreeTextView = songView.findViewById(R.id.nandu_2)
            rightHandDegreeRatingBar = songView.findViewById(R.id.nandu)
            leftHandDegreeRatingBar = songView.findViewById(R.id.leftnandu)
            songNameScrollText = songView.findViewById(R.id.s_n)
        }
    }
}